Add/rest pre insert filter terms

kjbenk opened the pull request:
Fixes this issue https://github.com/WP-API/WP-API/issues/2368

kjbenk committed:
Merge pull request #1 from WP-API/develop

Sync Repo

kjbenk committed:
Merge remote-tracking branch 'WP-API/develop' into develop

kjbenk committed:
Merge remote-tracking branch 'origin/develop' into develop

kjbenk committed:
Added rest_pre_insert filter to the terms controller

danielbachhuber commented on lib/endpoints/class-wp-rest-terms-controller.php:
```diff
@@ -528,6 +489,52 @@ public function delete_item( $request ) {
+
+		if ( isset( $request['parent'] ) ) {
+			if ( ! is_taxonomy_hierarchical( $this->taxonomy ) ) {
+				return new WP_Error( 'rest_taxonomy_not_hierarchical', __( 'Can not set resource parent, taxonomy is not hierarchical.' ), array( 'status' => 400 ) );
```
I don't think `prepare_item_for_database()` should ever return a `WP_Error`. You can silently ignore the `parent` param if it's not set in the schema.

kjbenk replied:
Would silently ignoring the `parent` parma be setting it to `-1` instead of throwing an error, or not setting it at all within the $prepared_term object?

Also, if I remove the returned WP_Error objects then should I remove these checks as well?
- https://github.com/WP-API/WP-API/pull/2377/files#diff-6a5f19de0f37a384ee52877de312ee67R409
- https://github.com/WP-API/WP-API/pull/2377/files#diff-6a5f19de0f37a384ee52877de312ee67R337

danielbachhuber replied:
> Would silently ignoring the parent parma be setting it to -1 instead of throwing an error, or not setting it at all within the $prepared_term object?

Probably setting it to 0, as that's the closest representation to a WP Term object

> Also, if I remove the returned WP_Error objects then should I remove these checks as well?

Yes.

danielbachhuber commented on lib/endpoints/class-wp-rest-terms-controller.php:
```diff
@@ -528,6 +489,52 @@ public function delete_item( $request ) {
+
+		if ( isset( $request['name'] ) ) {
+			$prepared_term->name = $request['name'];
+			$prepared_term->args['name'] = $request['name'];
```
Er, the `$prepared_term` object should behave similarly to a WP_Term object. Its attributes shouldn't be nested under `$args`.

kjbenk replied:
I see what you mean, but if I return a single term object then would that still work with the `wp_insert_term` function?  I see that you need to supply both the name and the args separately which is why I nested them that way.

https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-terms-controller.php#L360

Do you think a better way would be to return a WP_Term object and then supply the entire object as the args and then just use the name attribute for the first parameter of `wp_insert_term`?

Maybe like this `$term = wp_insert_term( $prepared_term->name, $this->taxonomy, $prepared_term );`

danielbachhuber replied:
> but if I return a single term object then would that still work with the wp_insert_term function? I

No, you'll need to break them apart.

> Do you think a better way would be to return a WP_Term object and then supply the entire object as the args and then just use the name attribute for the first parameter of wp_insert_term?

Yes, which is why I left the comment.

kjbenk committed:
Fixed Bugs

* FIXED: The prepare_item_for_database function will now only return a
valid WP_Term object

kjbenk commented:
@danielbachhuber let if me if that new commit https://github.com/WP-API/WP-API/pull/2377/commits/5bc83097db06f53c07db5bd8de4136f2904d0db7 is good for a merge.

danielbachhuber commented on lib/endpoints/class-wp-rest-terms-controller.php:
```diff
@@ -528,6 +505,63 @@ public function delete_item( $request ) {
+			$prepared_term->term_group = (int) $request['term_group'];
+		}
+
+		if ( isset( $request['term_taxonomy_id'] ) ) {
```
We don't support setting `term_taxonomy_id` through the REST API. `term_taxonomy_id` and `term_id` are the same in our eyes, and `term_id` is always exposed as `id`.

danielbachhuber commented on lib/endpoints/class-wp-rest-terms-controller.php:
```diff
@@ -528,6 +505,63 @@ public function delete_item( $request ) {
+			$prepared_term->parent = $parent_term_id;
+		}
+
+		if ( isset( $request['count'] ) ) {
```
`count` is a readonly field, and can't be set.

danielbachhuber commented on lib/endpoints/class-wp-rest-terms-controller.php:
```diff
@@ -528,6 +505,63 @@ public function delete_item( $request ) {
+		 * @param object          $prepared_term Term object.
+		 * @param WP_REST_Request $request       Request object.
+		 */
+		return apply_filters( 'rest_pre_insert_term', $prepared_term, $request );
```
This should be `rest_pre_insert_{$this->taxonomy}`

danielbachhuber commented on lib/endpoints/class-wp-rest-terms-controller.php:
```diff
@@ -528,6 +505,63 @@ public function delete_item( $request ) {
+			$prepared_term->slug = $request['slug'];
+		}
+
+		if ( isset( $request['term_group'] ) ) {
```
We don't permit setting the `term_group` through the API.

danielbachhuber commented:
@kjbenk Better. Left comments on a few more things needing to be changed.

kjbenk committed:
Fixed Bugs

* FIXED: Removed all term attributes that are either readonly or
irrelevant within the `prepare_item_for_database` function
* FIXED: Changes the `rest_pre_insert_` filter to account for all
taxonomies not just the generic `term`, so it is now
`rest_pre_insert_{$this->taxonomy}`

kjbenk commented:
@danielbachhuber thanks for the feedback, I am learning a lot :) The next commit has passed so you are free to review it.

danielbachhuber commented:
:ship: :ship: :ship:
